immortal-cravings.lua

```diff
@@ -0,0 +1,231 @@
+                dclosest = ditem
+            end
+        end
+    end
+    return closest
+end
+
+---find a drink
+---@param pos df.coord
+---@return df.item_drinkst|nil
+local function get_closest_drink(pos)
+    local is_good = function (drink)
+        local container = dfhack.items.getContainer(drink)
+        return container and df.item_barrelst:is_instance(container)
+    end
+    return findClosest(pos, df.global.world.items.other.DRINK, is_good)
+end
+
+---find some prepared meal
+---@return df.item_foodst?
+local function get_closest_meal(pos)
+    ---@param meal df.item_foodst
+    local function is_good(meal)
+        return meal.flags.rotten == false
+    end
+    return findClosest(pos, df.global.world.items.other.FOOD, is_good)
+end
+
+---create a Drink job for the given unit
+---@param unit df.unit
+local function goDrink(unit)
+    local drink = get_closest_drink(unit.pos)
+    if not drink then
+        -- print('no accessible drink found')
+        return
+    end
+    local job = idle.make_job()
+    job.job_type = df.job_type.DrinkItem
+    job.flags.special = true
+    local dx, dy, dz = dfhack.items.getPosition(drink)
+    job.pos = xyz2pos(dx, dy, dz)
+    if not dfhack.job.attachJobItem(job, drink, df.job_item_ref.T_role.Other, -1, -1) then
+        error('could not attach drink')
+        return
+    end
+    dfhack.job.addWorker(job, unit)
+    local name = dfhack.TranslateName(dfhack.units.getVisibleName(unit))
+    print(dfhack.df2console('immortal-cravings: %s is getting a drink'):format(name))
+end
+
+---create Eat job for the given unit
+---@param unit df.unit
+local function goEat(unit)
+    local meal = get_closest_meal(unit.pos)
+    if not meal then
+        -- print('no accessible meals found')
+        return
+    end
+    local job = idle.make_job()
+    job.job_type = df.job_type.Eat
+    job.flags.special = true
+    local dx, dy, dz = dfhack.items.getPosition(meal)
+    job.pos = xyz2pos(dx, dy, dz)
+    if not dfhack.job.attachJobItem(job, meal, df.job_item_ref.T_role.Other, -1, -1) then
+        error('could not attach meal')
+        return
+    end
+    dfhack.job.addWorker(job, unit)
+    local name = dfhack.TranslateName(dfhack.units.getVisibleName(unit))
+    print(dfhack.df2console('immortal-cravings: %s is getting something to eat'):format(name))
+end
+
+--- script logic
+
+local GLOBAL_KEY = 'immortal-cravings'
+
+enabled = enabled or false
+function isEnabled()
+    return enabled
+end
+
+local function persist_state()
+    dfhack.persistent.saveSiteData(GLOBAL_KEY, {
+        enabled=enabled,
+    })
+end
+
+--- Load the saved state of the script
+local function load_state()
+    -- load persistent data
+    local persisted_data = dfhack.persistent.getSiteData(GLOBAL_KEY, {})
+    enabled = persisted_data.enabled or false
+end
+
+DrinkAlcohol = df.need_type['DrinkAlcohol']
+EatGoodMeal = df.need_type['EatGoodMeal']
+
+---@type integer[]
+watched = {}
+
+threshold = -9000
```

Comment:
these should be local or "protected" globals (`foo = foo or default_foo`). otherwise, they'll get reset whenever the player runs (probably mistakenly) `immortal-cravings`

Reply:
`threshold` should definitely be `local` since it's read only. `watched` could be either